pkgs/racket-doc/scribblings/reference/eval-model.scrbl

```diff
@@ -60,55 +59,57 @@ Some simplifications require more than one step. For example:
 ]
 
 An expression that is not a @tech{value} can always be partitioned
-into two parts: a @deftech{redex}, which is the part that changed in a
+into two parts: a @deftech{redex} (``reducible expression''),
+which is the part that can change in a
 single-step simplification (highlighted), and the
 @deftech{continuation}, which is the evaluation
-context surrounding an expression. In @racket[(- 4 (+ 1 1))], the redex is @racket[(+ 1 1)], and
-the continuation is @racket[(- 4 @#,hole)], where @hole takes the
-place of the redex. That is, the continuation says how to ``continue''
+context surrounding the redex. In @racket[(- 4 (+ 1 1))], the redex is @racket[(+ 1 1)], and
+the continuation is @racket[(- 4 @#,hole)], where @hole indicates the
+position of the redex. That is, the continuation says how to ``continue''
```

Comment:
I don't see why we want to say that the hole "takes the place of" or "replaces" the redex. It's just an indicator in the continuation for documentation purposes. How about "represents the redex"?

Reply:
The idea here is that we can view Racket’s evaluation model as a series of syntactic transformations on expressions. So rather than imagining computation by visualizing an abstract machine with a program counter that executes instructions, we can literally reason about the behavior of programs via the manipulation of symbols. We can think of evaluating a program as a trace of reduction steps, so the program `(+ 1 (* 2 3))` would have the following trace:

`(+ 1 (* 2 3))` → `(+ 1 6)` → `7`

Note that this really can be a full specification of the system: we can think about the entire evaluation model as a series of purely syntactic transformations. However, in order to reason about evaluation order syntactically, we have to introduce this notion of continuations and redexes, which allow us to break expressions apart in well-defined ways. If we made the continuations and redexes explicit in the above trace, we’d have something like this instead:

`(+ 1 (* 2 3))` →<sub>decompose</sub> `(+ 1 [])`[`(* 2 3)`] →<sub>reduce</sub> `(+ 1 [])`[`6`] →<sub>recompose</sub> `(+ 1 6)` →<sub>decompose</sub> `[]`[`(+ 1 6)`] →<sub>reduce</sub> `[]`[`7`] →<sub>recompose</sub> `7`

What I’m getting at here is that the hole, `[]`, is *not* just an indication for documentation purposes! It’s a syntactic marker used in a specific, formal way to provide a syntactic reduction semantics for Racket. If you want to know more about reduction semantics, check out [PLT Redex](https://redex.racket-lang.org) and its accompanying book, [Semantics Engineering with PLT Redex](https://mitpress.mit.edu/books/semantics-engineering-plt-redex), which includes an explanation of reduction semantics and how it works.

Now, admittedly, in practice, Racket is not fully-specified this way. The evaluation model section of the Racket documentation does not include a formal treatment of Racket’s evaluation order in terms of evaluation contexts, nor does it provide a set of syntactic reduction rules. Maybe such a thing could be valuable, I don’t know, but the evaluation model section is somewhere in between: it’s not a formal specification, but it’s not entirely informal, either. In any case, I believe the takeaway is that this decomposition of a term into a redex and its continuation is a formal idea described informally, and a key idea of that decomposition is that it is syntactic; **the hole really does syntactically take the place of the redex** in the syntax we use to describe a continuation.

Reply:
> Anyway, what do you suggest instead of "indicates the position of". I think "takes the place of" suggests that the redex will never be recomposed into the expression.

That’s a fair point. Could you just clarify that by saying “@hole replaces the redex during its simplification”? I don’t like “indicates the position of” because it seems needlessly vague compared to “replaces”. It makes me wonder *how* it indicates the redex’s position, and the answer to that is both simple and precise: it replaces it, syntactically.

Reply:
Can we make it clearer by adding a few words? I'm only trying to make it clear that it does not take the place of the redex "permanently."

"... where [] takes the place of the redex as the redex is reduced."